Add unit-tests target

nilsdeppe opened the pull request:
## Proposed changes

Adds a `unit-tests` target that builds everything needed to run `ctest -L unit`.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add unit-tests target

nilsdeppe committed:
Switch algorithm tests to unit tests

nilsdeppe committed:
Add python tests to unit-tests target

kidder reviewed (dismissed):
@wthrowe do you want to look at this PR?

nilsdeppe commented:
One thought that just occurred to me: I need to re-order the `tests.yaml` so that we build `unit-test`, then run `ctest -L unit...`, then build the executables, then run `ctest -LE unit...`

Edit: I want CI to make sure the `unit-tests` target actually works :P

nilsdeppe committed:
Use unit-tests target in CI

Build the unit-tests target and then run `ctest -L unit` before building
test-executables to test that all unit tests are correctly built as part of
the unit-tests target.

nilsvu commented:
@kidder please take another look at this
